File: Tutorials/WinForms/DevExpress.MVVM/CS/Views/ViewID.cs
using System;
using DevExpress.Utils;

namespace WinFormsApplication.ViewModels {
    public sealed class ViewID : IEquatable<ViewID> {
        readonly object viewType;
        readonly int? id;
        public ViewID(object viewType) {
            this.viewType = viewType;
            this.id = null;
        }
        public ViewID(object viewType, int id) {
            this.viewType = viewType;
            this.id = id;
        }
        public bool Equals(ViewID other) {
            return (other != null) && Equals(viewType, other.viewType) && Equals(id, other.id);
        }
        public override bool Equals(object obj) {
            return Equals(obj as ViewID);
        }
        public override int GetHashCode() {
            return HashCodeHelper.CalculateGeneric(viewType, id);
        }
    }
}

File: Tutorials/WinForms/DevExpress.MVVM/VB/Views/ViewID.vb
Imports System
Imports DevExpress.Utils

Namespace WinFormsApplication.ViewModels
	Public NotInheritable Class ViewID
		Implements IEquatable(Of ViewID)

		Private ReadOnly viewType As Object
		Private ReadOnly id? As Integer
		Public Sub New(ByVal viewType As Object)
			Me.viewType = viewType
			Me.id = Nothing
		End Sub
		Public Sub New(ByVal viewType As Object, ByVal id As Integer)
			Me.viewType = viewType
			Me.id = id
		End Sub
		Public Overloads Function Equals(ByVal other As ViewID) As Boolean Implements IEquatable(Of ViewID).Equals
			Return (other IsNot Nothing) AndAlso Equals(viewType, other.viewType) AndAlso Equals(id, other.id)
		End Function
		Public Overrides Function Equals(ByVal obj As Object) As Boolean
			Return Equals(TryCast(obj, ViewID))
		End Function
		Public Overrides Function GetHashCode() As Integer
			Return HashCodeHelper.CalculateGeneric(viewType, id)
		End Function
	End Class
End Namespace
